crypto/dilithium/p_dilithium3.c

```diff
@@ -9,117 +9,300 @@
     return 0;
+      }
+
+  pkey->pkey.pqdsa_key = key;
+  return 1;
+}
+
+static int pkey_pqdsa_sign_message(EVP_PKEY_CTX *ctx, uint8_t *sig,
+                                        size_t *siglen, const uint8_t *tbs,
+                                        size_t tbslen) {
+  PQDSA_PKEY_CTX *dctx = ctx->data;
+  const PQDSA *pqdsa = dctx->pqdsa;
+  if (pqdsa == NULL) {
+    if (ctx->pkey == NULL) {
+      OPENSSL_PUT_ERROR(EVP, EVP_R_NO_PARAMETERS_SET);
+      return 0;
+    }
+    pqdsa = PQDSA_KEY_get0_sig(ctx->pkey->pkey.pqdsa_key);
   }
+
   // Caller is getting parameter values.
   if (sig == NULL) {
-    *siglen = DILITHIUM3_SIGNATURE_BYTES;
+    *siglen = pqdsa->signature_len;
     return 1;
   }
 
-  if (*siglen < DILITHIUM3_SIGNATURE_BYTES) {
+  if (*siglen < pqdsa->signature_len) {
     OPENSSL_PUT_ERROR(EVP, EVP_R_BUFFER_TOO_SMALL);
     return 0;
   }
 
-  if (ml_dsa_65_sign(sig, siglen, tbs, tbslen, NULL, 0, key->priv) != 0) {
+  // Check that the context is properly configured.
+  if (ctx->pkey == NULL ||
+      ctx->pkey->pkey.pqdsa_key == NULL ||
+      ctx->pkey->type != EVP_PKEY_NISTDSA) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_OPERATON_NOT_INITIALIZED);
+    return 0;
+      }
+
+  PQDSA_KEY *key = ctx->pkey->pkey.pqdsa_key;
+  if (!key->secret_key) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_NOT_A_PRIVATE_KEY);
+    return 0;
+  }
+
+  if (pqdsa->method->sign(sig, siglen, tbs, tbslen, NULL,
+                            0, key->secret_key) != 0) {
     OPENSSL_PUT_ERROR(EVP, ERR_R_INTERNAL_ERROR);
     return 0;
   }
   return 1;
 }
 
-static int pkey_dilithium3_verify_message(EVP_PKEY_CTX *ctx, const uint8_t *sig,
+static int pkey_pqdsa_verify_message(EVP_PKEY_CTX *ctx, const uint8_t *sig,
                                           size_t siglen, const uint8_t *tbs,
                                           size_t tbslen) {
-  if (ctx == NULL || ctx->pkey->pkey.ptr == NULL) {
-    OPENSSL_PUT_ERROR(EVP, EVP_R_MISSING_PARAMETERS);
-    return 0;
+  PQDSA_PKEY_CTX *dctx = ctx->data;
+  const PQDSA *pqdsa = dctx->pqdsa;
+
+  if (pqdsa == NULL) {
+    if (ctx->pkey == NULL) {
+      OPENSSL_PUT_ERROR(EVP, EVP_R_NO_PARAMETERS_SET);
+      return 0;
+    }
+
+    pqdsa = PQDSA_KEY_get0_sig(ctx->pkey->pkey.pqdsa_key);
   }
+  // Check that the context is properly configured.
+  if (ctx->pkey == NULL ||
+    ctx->pkey->pkey.pqdsa_key == NULL ||
+    ctx->pkey->type != EVP_PKEY_NISTDSA) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_OPERATON_NOT_INITIALIZED);
+    return 0;
+    }
 
-  DILITHIUM3_KEY *key = ctx->pkey->pkey.ptr;
+  PQDSA_KEY *key = ctx->pkey->pkey.pqdsa_key;
 
-  if (siglen != DILITHIUM3_SIGNATURE_BYTES ||
-      ml_dsa_65_verify(tbs, tbslen, sig, siglen, NULL, 0, key->pub) != 0) {
+  if (siglen != pqdsa->signature_len ||
+      pqdsa->method->verify(tbs, tbslen, sig, siglen,
+                              NULL, 0, key->public_key) != 0) {
     OPENSSL_PUT_ERROR(EVP, EVP_R_INVALID_SIGNATURE);
+    return 0;
+      }
```

Comment:
```suggestion
  }
```

Reply:
fixed in [e2234e3](https://github.com/aws/aws-lc/pull/1963/commits/e2234e347025d64cd9f912fb78555585e03a6198)